File: tests/bench/python/runner.py
#!/usr/bin/env sage
"""
Benchmark runner for SageMath.

Usage:
    sage tests/bench/python/runner.py --input tests/bench/inputs/arith.bench.inputs.json
    cat tests/bench/inputs/arith.bench.inputs.json | sage tests/bench/python/runner.py
"""

import json
import sys
import time
from sage.all import *
from sage.env import SAGE_VERSION


def hex_to_int(value):
    return int(value.replace(" ", "").replace("\n", ""), 16)


# Curve parameters (prime field short Weierstrass form)
SECP256K1 = {
    "p": hex_to_int("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F"),
    "a": 0,
    "b": 7,
    "gx": hex_to_int("79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798"),
    "gy": hex_to_int("483ADA7726A3C4655DA4FBFC0E1108A8FD17B448A68554199C47D08FFB10D4B8"),
}

P256 = {
    "p": hex_to_int("FFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFF"),
    "a": hex_to_int("FFFFFFFF00000001000000000000000000000000FFFFFFFFFFFFFFFFFFFFFFFC"),
    "b": hex_to_int("5AC635D8AA3A93E7B3EBBD55769886BC651D06B0CC53B0F63BCE3C3E27D2604B"),
    "gx": hex_to_int("6B17D1F2E12C4247F8BCE6E563A440F277037D812DEB33A0F4A13945D898C296"),
    "gy": hex_to_int("4FE342E2FE1A7F9B8EE7EB4A7C0F9E162BCE33576B315ECECBB6406837BF51F5"),
}

BLS12_381_G1 = {
    "p": hex_to_int(
        "1A0111EA397FE69A4B1BA7B6434BACD764774B84F38512BF6730D2A0F6B0F624"
        "1EABFFFEB153FFFFB9FEFFFFFFFFAAAB"
    ),
    "a": 0,
    "b": 4,
    "gx": int(
        "3685416753713387016781088315183077757961620795782546409894578378688607592378376318836054947676345821548104185464507"
    ),
    "gy": int(
        "1339506544944476473020471379941921221584933875938349620426543736416511423956333506472724655353366534992391756441569"
    ),
}

SECP256K1_FIELD = GF(SECP256K1["p"])
SECP256K1_CURVE = EllipticCurve(SECP256K1_FIELD, [SECP256K1["a"], SECP256K1["b"]])
SECP256K1_G = SECP256K1_CURVE([SECP256K1["gx"], SECP256K1["gy"]])

P256_FIELD = GF(P256["p"])
P256_CURVE = EllipticCurve(P256_FIELD, [P256["a"], P256["b"]])
P256_G = P256_CURVE([P256["gx"], P256["gy"]])

BLS12_381_FIELD = GF(BLS12_381_G1["p"])
BLS12_381_CURVE = EllipticCurve(BLS12_381_FIELD, [BLS12_381_G1["a"], BLS12_381_G1["b"]])
BLS12_381_G1_POINT = BLS12_381_CURVE([BLS12_381_G1["gx"], BLS12_381_G1["gy"]])


def parse_arg(arg):
    if isinstance(arg, list):
        return [Integer(value) for value in arg]
    return Integer(arg)


def percentile(values, p):
    if not values:
        return None
    values = sorted(values)
    index = (len(values) - 1) * p
    lower = int(index)
    upper = int(index + 1)
    if upper >= len(values):
        return values[lower]
    weight = index - lower
    return values[lower] * (1 - weight) + values[upper] * weight


def execute_function(module, function_name, args):
    function_map = {
        'arith': {
            'gcd': gcd,
            'lcm': lcm,
            'xgcd': xgcd,
            'factor': factor,
            'is_prime': is_prime,
            'is_prime_power': is_prime_power,
            'next_prime': next_prime,
            'previous_prime': previous_prime,
            'euler_phi': euler_phi,
            'radical': radical,
            'moebius': moebius,
            'kronecker_symbol': kronecker_symbol,
            'legendre_symbol': legendre_symbol,
            'jacobi_symbol': jacobi_symbol,
            'power_mod': power_mod,
            'inverse_mod': inverse_mod,
            'crt': crt,
            'isqrt': isqrt,
            'is_square': is_square,
            'is_squarefree': is_squarefree,
            'divisors': divisors,
            'number_of_divisors': number_of_divisors,
            'sigma': sigma,
            'prime_range': prime_range,
            'trial_division': trial_division,
            'squarefree_part': squarefree_part,
            'prime_factors': prime_factors,
            'valuation': valuation,
        },
        'bench_crypto': {
            'rsa_2048_pow': lambda m, e, n: power_mod(m % n, e, n),
            'secp256k1_mul': lambda k: SECP256K1_G * k,
            'p256_mul': lambda k: P256_G * k,
            'bls12_381_g1_mul': lambda k: BLS12_381_G1_POINT * k,
        },
    }

    if module not in function_map:
        raise ValueError(f"Unknown module: {module}")

    if function_name not in function_map[module]:
        raise ValueError(f"Unknown function: {module}.{function_name}")

    func = function_map[module][function_name]
    return func(*args)


def run_case(module_name, bench_case):
    inputs = [[parse_arg(arg) for arg in args] for args in bench_case['inputs']]
    warmup = int(bench_case.get('warmup', 0))
    iterations = int(bench_case.get('iterations', 1))
    repeats = int(bench_case.get('repeats', 1))

    total_ns = []
    per_call_ns = []
    error = None

    try:
        for i in range(warmup):
            args = inputs[i % len(inputs)]
            execute_function(module_name, bench_case['function'], args)

        for _ in range(repeats):
            start = time.perf_counter_ns()
            for i in range(iterations):
                args = inputs[i % len(inputs)]
                execute_function(module_name, bench_case['function'], args)
            end = time.perf_counter_ns()
            elapsed = end - start
            total_ns.append(str(elapsed))
            per_call_ns.append(elapsed / iterations)
    except Exception as exc:
        error = str(exc)
        total_ns = []
        per_call_ns = []

    return {
        'id': bench_case.get('id', bench_case['function']),
        'function': bench_case['function'],
        'inputs': len(inputs),
        'iterations': iterations,
        'warmup': warmup,
        'repeats': repeats,
        'total_ns': total_ns,
        'per_call_ns': per_call_ns,
        'p50_ns': percentile(per_call_ns, 0.5),
        'p90_ns': percentile(per_call_ns, 0.9),
        'error': error,
    }


def read_input(input_path):
    if input_path:
        with open(input_path, 'r', encoding='utf-8') as handle:
            return handle.read()
    return sys.stdin.read()


def main():
    input_path = None
    args = sys.argv[1:]
    for i, arg in enumerate(args):
        if arg == '--input' and i + 1 < len(args):
            input_path = args[i + 1]

    raw = read_input(input_path)
    if not raw.strip():
        print('Error: No input provided. Pass benchmark inputs via --input or stdin.', file=sys.stderr)
        sys.exit(1)

    suite = json.loads(raw)
    results = [run_case(suite['module'], bench_case) for bench_case in suite['cases']]

    output = {
        'module': suite['module'],
        'inputFile': input_path,
        'runtime': f"sage {SAGE_VERSION}",
        'timestamp': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
        'results': results,
    }

    print(json.dumps(output, indent=2))


if __name__ == '__main__':
    main()

File: tests/bench/typescript/runner.ts
#!/usr/bin/env bun
/**
 * Benchmark runner for sagemath-ts.
 *
 * Usage:
 *   bun tests/bench/typescript/runner.ts --input tests/bench/inputs/arith.bench.inputs.json
 *   cat tests/bench/inputs/arith.bench.inputs.json | bun tests/bench/typescript/runner.ts
 */

import {
  gcd,
  lcm,
  xgcd,
  factor,
  is_prime,
  is_prime_power,
  next_prime,
  previous_prime,
  euler_phi,
  radical,
  moebius,
  kronecker_symbol,
  legendre_symbol,
  jacobi_symbol,
  power_mod,
  inverse_mod,
  crt,
  isqrt,
  is_square,
  is_squarefree,
  divisors,
  number_of_divisors,
  sigma,
  prime_range,
  squarefree_part,
  prime_factors,
  valuation,
  arith,
} from '../../../packages/sagemath-ts/src/index.js';
import { readFileSync } from 'node:fs';
import {
  bls12_381_g1,
  createPrimeFieldCurve,
  p256,
  secp256k1,
} from '../../../packages/zksecurity-cheatsheets/src/index.js';

interface BenchCaseInputs {
  id: string;
  function: string;
  inputs: (string | string[])[][];
  warmup: number;
  iterations: number;
  repeats: number;
}

interface BenchSuiteInputs {
  module: string;
  generatedAt: string;
  sourceFile?: string;
  cases: BenchCaseInputs[];
}

interface BenchCaseResult {
  id: string;
  function: string;
  inputs: number;
  iterations: number;
  warmup: number;
  repeats: number;
  total_ns: string[];
  per_call_ns: number[];
  p50_ns: number | null;
  p90_ns: number | null;
  error: string | null;
}

interface BenchSuiteResult {
  module: string;
  inputFile: string | null;
  runtime: string;
  timestamp: string;
  results: BenchCaseResult[];
}

const secp256k1Context = createPrimeFieldCurve(secp256k1);
const p256Context = createPrimeFieldCurve(p256);
const bls12_381_g1Context = createPrimeFieldCurve(bls12_381_g1);

function parseArg(arg: string | string[]): bigint | bigint[] {
  if (Array.isArray(arg)) {
    return arg.map((value) => BigInt(value));
  }
  return BigInt(arg);
}

function percentile(values: number[], p: number): number | null {
  if (values.length === 0) {
    return null;
  }
  const sorted = [...values].sort((a, b) => a - b);
  const index = (sorted.length - 1) * p;
  const lower = Math.floor(index);
  const upper = Math.ceil(index);
  if (lower === upper) {
    return sorted[lower]!;
  }
  const weight = index - lower;
  return sorted[lower]! * (1 - weight) + sorted[upper]! * weight;
}

function executeFunction(
  module: string,
  functionName: string,
  args: (bigint | bigint[])[]
): unknown {
  const functionMap: Record<string, Record<string, (...args: any[]) => unknown>> = {
    arith: {
      gcd: (a: bigint, b: bigint) => gcd(a, b),
      lcm: (a: bigint, b: bigint) => lcm(a, b),
      xgcd: (a: bigint, b: bigint) => xgcd(a, b),
      factor: (n: bigint) => factor(n),
      is_prime: (n: bigint) => is_prime(n),
      is_prime_power: (n: bigint) => is_prime_power(n),
      next_prime: (n: bigint) => next_prime(n),
      previous_prime: (n: bigint) => previous_prime(n),
      euler_phi: (n: bigint) => euler_phi(n),
      radical: (n: bigint) => radical(n),
      moebius: (n: bigint) => moebius(n),
      kronecker_symbol: (a: bigint, n: bigint) => kronecker_symbol(a, n),
      legendre_symbol: (a: bigint, p: bigint) => legendre_symbol(a, p),
      jacobi_symbol: (a: bigint, n: bigint) => jacobi_symbol(a, n),
      power_mod: (a: bigint, n: bigint, m: bigint) => power_mod(a, n, m),
      inverse_mod: (a: bigint, m: bigint) => inverse_mod(a, m),
      crt: (a: bigint, b: bigint, m: bigint, n: bigint) => crt(a, b, m, n),
      isqrt: (n: bigint) => isqrt(n),
      is_square: (n: bigint) => is_square(n),
      is_squarefree: (n: bigint) => is_squarefree(n),
      divisors: (n: bigint) => divisors(n),
      number_of_divisors: (n: bigint) => number_of_divisors(n),
      sigma: (n: bigint, k?: bigint) => sigma(n, k),
      prime_range: (start: bigint, stop?: bigint) =>
        stop !== undefined ? prime_range(start, stop) : prime_range(start),
      trial_division: (n: bigint, bound?: bigint) =>
        bound !== undefined ? arith.trial_division(n, bound) : arith.trial_division(n),
      squarefree_part: (n: bigint) => squarefree_part(n),
      prime_factors: (n: bigint) => prime_factors(n),
      valuation: (n: bigint, p: bigint) => valuation(n, p),
    },
    bench_crypto: {
      rsa_2048_pow: (m: bigint, e: bigint, n: bigint) => power_mod(m % n, e, n),
      secp256k1_mul: (k: bigint) => secp256k1Context.generator.mul(k),
      p256_mul: (k: bigint) => p256Context.generator.mul(k),
      bls12_381_g1_mul: (k: bigint) => bls12_381_g1Context.generator.mul(k),
    },
  };

  if (!(module in functionMap)) {
    throw new Error(`Unknown module: ${module}`);
  }

  if (!(functionName in functionMap[module]!)) {
    throw new Error(`Unknown function: ${module}.${functionName}`);
  }

  const func = functionMap[module]![functionName]!;
  return func(...args);
}

function runCase(moduleName: string, benchCase: BenchCaseInputs): BenchCaseResult {
  const inputs = benchCase.inputs.map((args) => args.map((arg) => parseArg(arg)));
  const warmup = benchCase.warmup ?? 0;
  const iterations = benchCase.iterations ?? 1;
  const repeats = benchCase.repeats ?? 1;

  let error: string | null = null;
  let totalNs: string[] = [];
  let perCallNs: number[] = [];

  try {
    for (let i = 0; i < warmup; i++) {
      const args = inputs[i % inputs.length]!;
      executeFunction(moduleName, benchCase.function, args);
    }

    for (let r = 0; r < repeats; r++) {
      const start = process.hrtime.bigint();
      for (let i = 0; i < iterations; i++) {
        const args = inputs[i % inputs.length]!;
        executeFunction(moduleName, benchCase.function, args);
      }
      const end = process.hrtime.bigint();
      const elapsed = end - start;
      totalNs.push(elapsed.toString());
      perCallNs.push(Number(elapsed) / iterations);
    }
  } catch (e) {
    error = e instanceof Error ? e.message : String(e);
    totalNs = [];
    perCallNs = [];
  }

  return {
    id: benchCase.id,
    function: benchCase.function,
    inputs: inputs.length,
    iterations,
    warmup,
    repeats,
    total_ns: totalNs,
    per_call_ns: perCallNs,
    p50_ns: percentile(perCallNs, 0.5),
    p90_ns: percentile(perCallNs, 0.9),
    error,
  };
}

async function readInput(inputPath: string | null): Promise<string> {
  if (inputPath) {
    return readFileSync(inputPath, 'utf8');
  }

  const chunks: string[] = [];
  const reader = Bun.stdin.stream().getReader();

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    chunks.push(new TextDecoder().decode(value));
  }

  return chunks.join('');
}

function parseArgs() {
  const args = process.argv.slice(2);
  let inputPath: string | null = null;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--input' && i + 1 < args.length) {
      inputPath = args[i + 1]!;
      i++;
    }
  }

  return { inputPath };
}

async function main() {
  const { inputPath } = parseArgs();
  const inputData = await readInput(inputPath);

  if (!inputData.trim()) {
    console.error('Error: No input provided. Pass benchmark inputs via --input or stdin.');
    process.exit(1);
  }

  const suite = JSON.parse(inputData) as BenchSuiteInputs;
  const results = suite.cases.map((benchCase) => runCase(suite.module, benchCase));

  const output: BenchSuiteResult = {
    module: suite.module,
    inputFile: inputPath,
    runtime: `bun ${Bun.version}`,
    timestamp: new Date().toISOString(),
    results,
  };

  console.log(JSON.stringify(output, null, 2));
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});
